_posts/2018-01-30-quicli.md

```diff
@@ -0,0 +1,294 @@
+but we were both willing to join forces
+and try to make one great crate
+that takes the best of both approaches.
+Garrett opened a few issues on [quicli's issue tracker][issues] (thanks!),
+which basically concluded in [#19],
+which was originally titled
+"merge stdcli and this lib".
+
+[vitiral]: https://github.com/vitiral
+[replied]: https://www.reddit.com/r/rust/comments/7s3zsd/quicli_quickly_build_cool_cli_apps_in_rust/dt28i2b/
+[stdcli]: https://github.com/vitiral/stdcli
+[issues]: https://github.com/killercup/quicli/issues
+[#19]: https://github.com/killercup/quicli/issues/19
+
+That sounded pretty awesome and made me really proud:
+Did my library
+– which was barely a day old at this point –
+look cool enough to compete with their existing effort?
+
+But, before blindly agreeing to this,
+I needed to take a step back
+and evaluate what I want quicli to be.
+Because, let's be honest,
+I hadn't really thought about a 'grand quicli vision'
+before.
+And why should I?
+I just wanted a small framework-like thing
+that made writing CLI apps less of a pain.
+But how exactly do I want to to this?
+Thanks to Garrett's questions and comments
+I now have an answer to this.
+
+## What do I want to quicli to be
+
+(The following are basically quotes from [#19].)
+
+My goal is to provide
+an opinionated set
+of convenience functions
+to quickly build CLI apps.
+
+In its implementation,
+you get what I'd call
+"a small framework around your main function":
+quicli's `main!` macro set up some basic things,
+and its prelude gives a few tools
+that are commonly required in CLI apps.
+These tools are either re-exports from existing Rust crates,
+or abstractions around library functions that are more ergonomic
+that better fit the CLI use-case
+(as well as my ergonomic requirements).
+
+What I want you to end up with is code that is
+concise,
+well-structured and boilerplate-free,
+and production-ready.
+Okay, that's nothing new; we all want that.
+But what does this even mean for a CLI app?
+
+- concise
+
+	Ideally: You type one line to get one thing done.
+	If you want to read a file
+	and see if its text contains the word "foobar",
+	you can do:
+
+	```rust
+	let content = read_file("./input.txt")?;
+	let has_foobar = content.contains("foobar");
+	```
+	
+	You want to read a file, I give you a `read_file` function.
+	This skips allocating a String as buffer, opening File, and calling `read_to_string` on that file.
+	And while reading a file to a string this way may not be the most performant option,
+	it is very concise.
+	
+	Another aspect of conciseness:
+	Instead of using `.unwrap` to deal with errors,
+	quicli's `main!` allows you to use `?` to propagate errors,
+	and shows human-readable messages when your program exited with an error state.
+	Similarly, instead of
+	writing `return Err(SomeType::new("message"))`,
+	by using quicli you also get [failure]'s `bail!` and `ensure!`.
+	
+- well-structured and boilerplate-free
```

Comment:
I still think it should be capitalized or at least made more clear as a section rather than a list but that's personal preference. Your choice really.